miniapps/dpg/util/weakform.cpp

```diff
@@ -0,0 +1,844 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "weakform.hpp"
+
+namespace mfem
+{
+
+void DPGWeakForm::Init()
+{
+   trial_integs.SetSize(trial_fes.Size(), test_fecols.Size());
+   for (int i = 0; i < trial_integs.NumRows(); i++)
+   {
+      for (int j = 0; j < trial_integs.NumCols(); j++)
+      {
+         trial_integs(i,j) = new Array<BilinearFormIntegrator * >();
+      }
+   }
+
+   test_integs.SetSize(test_fecols.Size(), test_fecols.Size());
+   for (int i = 0; i < test_integs.NumRows(); i++)
+   {
+      for (int j = 0; j < test_integs.NumCols(); j++)
+      {
+         test_integs(i,j) = new Array<BilinearFormIntegrator * >();
+      }
+   }
+
+   lfis.SetSize(test_fecols.Size());
+   for (int j = 0; j < lfis.Size(); j++)
+   {
+      lfis[j] = new Array<LinearFormIntegrator * >();
+   }
+
+
+   ComputeOffsets();
+
+   mat = mat_e = NULL;
+   diag_policy = mfem::Operator::DIAG_ONE;
+   height = dof_offsets[nblocks];
+   width = height;
+
+   initialized = true;
+   static_cond = nullptr;
+
+   if (store_matrices)
+   {
+      Bmat.SetSize(mesh->GetNE());
+      fvec.SetSize(mesh->GetNE());
+   }
+}
+
+void DPGWeakForm::ComputeOffsets()
+{
+   dof_offsets.SetSize(nblocks+1);
+   tdof_offsets.SetSize(nblocks+1);
+   dof_offsets[0] = 0;
+   tdof_offsets[0] = 0;
+   for (int i =0; i<nblocks; i++)
+   {
+      dof_offsets[i+1] = trial_fes[i]->GetVSize();
+      tdof_offsets[i+1] = trial_fes[i]->GetTrueVSize();
+   }
+   dof_offsets.PartialSum();
+   tdof_offsets.PartialSum();
+}
+
+// Allocate SparseMatrix and RHS
+void DPGWeakForm::AllocMat()
+{
+   if (static_cond) { return; }
+
+   mat = new BlockMatrix(dof_offsets);
+   mat->owns_blocks = 1;
+
+   for (int i = 0; i<mat->NumRowBlocks(); i++)
+   {
+      int h = dof_offsets[i+1] - dof_offsets[i];
+      for (int j = 0; j<mat->NumColBlocks(); j++)
+      {
+         int w = dof_offsets[j+1] - dof_offsets[j];
+         mat->SetBlock(i,j,new SparseMatrix(h, w));
+      }
+   }
+   y = new BlockVector(dof_offsets);
+   *y = 0.;
+}
+
+void DPGWeakForm::Finalize(int skip_zeros)
+{
+   if (mat) { mat->Finalize(skip_zeros); }
+   if (mat_e) { mat_e->Finalize(skip_zeros); }
+   if (static_cond) { static_cond->Finalize(); }
+}
+
+/// Adds new Domain BF Integrator. Assumes ownership of @a bfi.
+void DPGWeakForm::AddTrialIntegrator(
+   BilinearFormIntegrator *bfi, int n, int m)
+{
+   MFEM_VERIFY(n<trial_fes.Size(),
```

Comment:
It wouldn't hurt to check `0 <= n` (similarly in other checks).

Reply:
done!